### ISSUE TITLE
SMTP configuration issue

### ISSUE THREAD

**Issue by benpro** [Chocobozzz/PeerTube#373]
Hi there, it seems peertube cannot connect to a local SMTP server (basic postfix) in plain text, he tries to use STARTTLS and fails.

* **PeerTube version or URL:** v1.0.0-beta.1
* **NodeJS version:** v8.10.0

* **What is the expected behaviour?**

Peertube can connect to the local SMTP server in plain text.

* **What do you see instead?**

Peertube can't connect to the local SMTP server in plain text.

* **Server log if useful (Gist/Pastebin...):**

Config: 

```
smtp:
  hostname: 'localhost'
  port: 25
  username: null
  password: null
  tls: false
  ca_file: null # Used for self signed certificates
  from_address: 'videos@benpro.fr'
```

Logs:

```
Mar 21 22:35:58 web-peertube peertube[16797]: npm ERR! code ELIFECYCLE
Mar 21 22:35:58 web-peertube peertube[16797]: }
Mar 21 22:35:58 web-peertube peertube[16797]:   }
Mar 21 22:35:58 web-peertube peertube[16797]:     "command": "CONN"
Mar 21 22:35:58 web-peertube peertube[16797]:     "code": "ESOCKET",
Mar 21 22:35:58 web-peertube peertube[16797]:   "meta": {
Mar 21 22:35:58 web-peertube peertube[16797]: [videos.benpro.fr:443] 2018-03-21 22:35:58.954 error: Failed to connect to SMTP localhost:25. {
Mar 21 22:35:58 web-peertube postfix/smtpd[15409]: disconnect from localhost[127.0.0.1] ehlo=1 starttls=1 commands=2
Mar 21 22:35:58 web-peertube postfix/smtpd[15409]: lost connection after STARTTLS from localhost[127.0.0.1]
Mar 21 22:35:58 web-peertube postfix/smtpd[15409]: connect from localhost[127.0.0.1]
```

TCPdump:

```
22:31:04.231036 IP (tos 0x0, ttl 64, id 16737, offset 0, flags [DF], proto TCP (6), length 117)
    localhost.smtp > localhost.50774: Flags [P.], cksum 0xfe69 (incorrect -> 0xd7db), seq 1:66, ack 1, win 342, options [nop,nop,TS val 1490769110 ecr 1490769110], length 65: SMTP, length: 65
        220 web-peertube.lxd.saber.benpro.fr ESMTP Postfix (Debian/GNU)
22:31:04.234946 IP (tos 0x0, ttl 64, id 34174, offset 0, flags [DF], proto TCP (6), length 70)
    localhost.50774 > localhost.smtp: Flags [P.], cksum 0xfe3a (incorrect -> 0x8bda), seq 1:19, ack 66, win 342, options [nop,nop,TS val 1490769111 ecr 1490769110], length 18: SMTP, length: 18
        EHLO [127.0.0.1]
22:31:04.234964 IP (tos 0x0, ttl 64, id 16738, offset 0, flags [DF], proto TCP (6), length 52)
    localhost.smtp > localhost.50774: Flags [.], cksum 0xfe28 (incorrect -> 0x44f8), ack 19, win 342, options [nop,nop,TS val 1490769111 ecr 1490769111], length 0
22:31:04.235041 IP (tos 0x0, ttl 64, id 16739, offset 0, flags [DF], proto TCP (6), length 221)
    localhost.smtp > localhost.50774: Flags [P.], cksum 0xfed1 (incorrect -> 0xd3f0), seq 66:235, ack 19, win 342, options [nop,nop,TS val 1490769111 ecr 1490769111], length 169: SMTP, length: 169
        250-web-peertube.lxd.saber.benpro.fr
        250-PIPELINING
        250-SIZE 10240000
        250-VRFY
        250-ETRN
        250-STARTTLS
        250-ENHANCEDSTATUSCODES
        250-8BITMIME
        250-DSN
        250 SMTPUTF8
22:31:04.235568 IP (tos 0x0, ttl 64, id 34175, offset 0, flags [DF], proto TCP (6), length 62)
    localhost.50774 > localhost.smtp: Flags [P.], cksum 0xfe32 (incorrect -> 0x01dd), seq 19:29, ack 235, win 350, options [nop,nop,TS val 1490769111 ecr 1490769111], length 10: SMTP, length: 10
        STARTTLS
22:31:04.235597 IP (tos 0x0, ttl 64, id 16740, offset 0, flags [DF], proto TCP (6), length 82)
    localhost.smtp > localhost.50774: Flags [P.], cksum 0xfe46 (incorrect -> 0x2cfc), seq 235:265, ack 29, win 342, options [nop,nop,TS val 1490769111 ecr 1490769111], length 30: SMTP, length: 30
        220 2.0.0 Ready to start TLS

```

**Comment by Chocobozzz** [comment]
If you don't want to use StartTLS, or if you use a self signed certificate maybe you could just disable StartTLS on Postfix?

**Comment by rhaamo** [comment]
postfix by default doesn't seems to have STARTTLS, only plain port 25.
the issue is that by using ```tls: false``` and ```port: 25```, peertube still force to use STARTTLS on port 25, which fails.

**Comment by Chocobozzz** [comment]
PeerTube only force StartTLS if the remote server supports it. You tcpdump seems to show that your Postfix supports StartTLS (BTW very nice bug report, thanks!).

```
22:31:04.235041 IP (tos 0x0, ttl 64, id 16739, offset 0, flags [DF], proto TCP (6), length 221)
    localhost.smtp > localhost.50774: Flags [P.], cksum 0xfed1 (incorrect -> 0xd3f0), seq 66:235, ack 19, win 342, options [nop,nop,TS val 1490769111 ecr 1490769111], length 169: SMTP, length: 169
        250-web-peertube.lxd.saber.benpro.fr
        250-PIPELINING
        250-SIZE 10240000
        250-VRFY
        250-ETRN
        250-STARTTLS
        250-ENHANCEDSTATUSCODES
        250-8BITMIME
        250-DSN
        250 SMTPUTF8
```

**Comment by benpro** [comment]
OK, I disabled TLS on my postfix configuration and it works.
But, this is the first I see a software fail like this.
IMO, Peertube should not try TLS if `tls: false` is setted...

You can close if you want.

**Comment by Chocobozzz** [comment]
`tls: false` is for TLS (port 465), not StartTLS (SMTP you know...). We could add a setting to force disabling StartTLS too but I'm not comfortable with that. And I don't see why people would not want to use StartTLS on a server that is configured to support it, or it's a misconfiguration.

**Comment by wxcafe** [comment]
using email with a self-signed certificate is very common and works perfectly well. Using StartTLS on localhost is generally not an issue because software will either a) detect it's localhost and go over plaintext, or b) ignore the self-signed certificate, as all email software does.

This is not an issue of misconfiguration.

**Comment by Chocobozzz** [comment]
Ignoring self-signed certificates is just a bad practice for an email client. And we provide a field in the configuration so people can add their ca file.

Forcing StartTLS to be disabled if the host is `localhost` is something we can do though

**Comment by ldidry** [comment]
@Chocobozzz you made a little confusion on nodemailer configuration.

According to https://nodemailer.com/smtp/:

- `tls` is for using [node.js TLSSocket options](https://nodejs.org/api/tls.html#tls_class_tls_tlssocket)
- `secure` is for using (or not) TLS
- If you want to ignore STARTTLS, set `secure` to false and `ignoreTLS` to true.

**Comment by benpro** [comment]
> We could add a setting to force disabling StartTLS too but I'm not comfortable with that. And I don't see why people would not want to use StartTLS on a server that is configured to support it, or it's a misconfiguration.

Well, expect many users to cry when encountering this same issue as me. Sysadmins using the local MTA listening on port 25 in plain text is pretty common.
At least, add a comment in the configuration file.

Oh and BTW, my postfix is working with a self-signed certificate (snakeoil) which is also pretty standard, so I don't known why Peertube fail to connect with STARTTLS. Maybe it wants a valid certificate delivered by an authority in ca-certificates? Need some more debug messages!

**Comment by Chocobozzz** [comment]
@ldidry `tls: false` is in the PeerTube config, I convert it to `secure: false`... Read the code https://github.com/Chocobozzz/PeerTube/blob/develop/server/lib/emailer.ts#L46

@benpro I don't like to see people crying. But if sysadmins configure their MTA to use a self signed certificate with STARTTLS enabled, they have to add their ca file yes. I added more logs this morning after seeing your issue https://github.com/Chocobozzz/PeerTube/commit/05e67d6206669b4a17e786038b1923e79bacb196

I'll force the deactivation of STARTTLS on `localhost` to avoid some future issues.

**Comment by benpro** [comment]
>  I added more logs this morning after seeing your issue 05e67d6

Good! Because when I need to use tcpdump to understand what the hell is happening, it annoys me...

 > I'll force the deactivation of STARTTLS on localhost to avoid some future issues.

I think this is the best compromise. I mean, plain text on localhost things is pretty standard (and hell yes, easy do debug).

**Comment by Chocobozzz** [comment]
:+1: 

See https://github.com/Chocobozzz/PeerTube/commit/376220e80e7f544808dce5324de676eed6dc6f43. Please let me know if it won't work as expected (available in the next release).

Thanks again for the issue, I know how it's frustrating when we have to use tcpdump :)

**Comment by wxcafe** [comment]
ignoring self-signed certificates is pretty common, and not bad practice at all, actually. A lot of software does it, including but not limited to postfix, exim, opensmtpd, [..]

**Comment by Aldarone** [comment]
Unfortunately this doesn't take care of other valid localhost addresses such as ::1/128 and 127.0.0.0/8 😕

**Comment by anmol26s** [comment]
How to send email for testing purpose?

**Comment by Chocobozzz** [comment]
PeerTube automatically test the SMTP connection at startup. But if you really want to test a real email, you can use "I forgot my password" in the login form.

**Comment by drzraf** [comment]
Stumbled upon this too.
TL;DR:

`config.smtp` value:
```
    secure: false
    ignoreTLS: true
    tls: false
```
(all 3 are needed)
